contracts/DivviRegistry.sol

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {AccessControlDefaultAdminRulesUpgradeable} from '@openzeppelin/contracts-upgradeable/access/extensions/AccessControlDefaultAdminRulesUpgradeable.sol';
+import {Initializable} from '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import {UUPSUpgradeable} from '@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol';
+
+/**
+ * @title DivviRegistry
+ * @notice A registry contract for managing Divvi entities and agreements
+ */
+contract DivviRegistry is
+  Initializable,
+  AccessControlDefaultAdminRulesUpgradeable,
+  UUPSUpgradeable
+{
+  struct EntityData {
+    bool exists;
+    bool requiresApproval;
+    // fields can be added here in a future upgrade if needed
+    // this is upgrade safe as long as `EntityData` is only used in a mapping
+  }
+
+  // Entities storage
+  mapping(address => EntityData) private _entities;
+
+  // Agreement storage
+  mapping(bytes32 => bool) private _agreements; // keccak256(provider, consumer) => true (if agreement exists)
+
+  // Events
+  event RewardsEntityRegistered(address indexed entity, bool requiresApproval);
+  event RequiresApprovalForRewardsAgreements(
+    address indexed entity,
+    bool requiresApproval
+  );
+  event RewardsAgreementRegistered(
+    address indexed rewardsProvider,
+    address indexed rewardsConsumer
+  );
+
+  // Errors
+  error EntityAlreadyExists(address entity);
+  error EntityDoesNotExist(address entity);
+  error AgreementAlreadyExists(address provider, address consumer);
+  error ProviderRequiresApproval(address provider);
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  /**
+   * @notice Initialize the contract with an owner and transfer delay
+   * @param owner The address that will have the DEFAULT_ADMIN_ROLE
+   * @param transferDelay The delay in seconds before admin role can be transferred
+   */
+  function initialize(address owner, uint48 transferDelay) public initializer {
+    __AccessControlDefaultAdminRules_init(transferDelay, owner);
+    __UUPSUpgradeable_init();
+  }
+
+  /**
+   * @notice Authorize contract upgrades
+   */
+  function _authorizeUpgrade(
+    address
+  ) internal override onlyRole(DEFAULT_ADMIN_ROLE) {} // solhint-disable-line no-empty-blocks
+
+  /**
+   * @notice Modifier to ensure an entity exists
+   * @param entity The entity address to check
+   */
+  modifier entityExists(address entity) {
+    if (!_entities[entity].exists) {
+      revert EntityDoesNotExist(entity);
+    }
+    _;
+  }
+
+  /**
+   * @notice Register the caller as a new rewards entity
+   * @param requiresApproval Whether the entity requires approval for agreements
+   */
+  function registerRewardsEntity(bool requiresApproval) external {
+    if (_entities[msg.sender].exists) {
```

Comment:
For `registerRewardsEntity` and `registerAgreementAsConsumer` can we utilize a signature instead of using msg.sender to get the entity's address?

This would allow us to build gas-less registration of entities which could be useful for an Ambassador program or an invite rewards scenario where the RewardsConsumer is just a person as opposed to an app.

cc @silasbw

Reply:
that has tradeoffs for the usability of these 2 methods though, because the signed message would need to be generated first and then passed as a method argument so entities can't just do the transaction straight from a block explorer / Safe UI? it seems like we should create new methods for those and restrict the caller to a known address instead?

Reply:
i propose that we address this requirement in a follow up PR (this is a new requirement that i am just hearing about now). this contract won't be deployed until we select a home network so there is no redeployment overhead to address this later

Reply:
Yes we can punt on this.

An an alternative to custom signed messages, we could adopt ERC-2771 for this.
This is what Gelato recommends: https://docs.gelato.network/web3-services/relay/erc-2771-recommended/sponsoredcallerc2771
We'd still need signing from the user and trusted relayer(s), but it is standard.

Reply:
okay, sounds good. Thanks 👍

Reply:
created this ticket https://linear.app/valora/issue/ENG-309/divviregistry-allow-registering-entities-on-their-behalf